Firestore/Source/API/FSTFirestoreComponent.mm

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#import "Firestore/Source/API/FSTFirestoreComponent.h"
+
+#import <FirebaseAuthInterop/FIRAuthInterop.h>
+#import <FirebaseCore/FIRAppInternal.h>
+#import <FirebaseCore/FIRComponent.h>
+#import <FirebaseCore/FIRComponentContainer.h>
+#import <FirebaseCore/FIRComponentRegistrant.h>
+#import <FirebaseCore/FIRDependency.h>
+#import <FirebaseCore/FIROptions.h>
+
+#include <memory>
+#include <utility>
+
+#import "Firestore/Source/API/FIRFirestore+Internal.h"
+#import "Firestore/Source/Util/FSTDispatchQueue.h"
+#import "Firestore/Source/Util/FSTUsageValidation.h"
+#include "Firestore/core/src/firebase/firestore/auth/credentials_provider.h"
+#include "Firestore/core/src/firebase/firestore/auth/firebase_credentials_provider_apple.h"
+#include "Firestore/core/src/firebase/firestore/util/hard_assert.h"
+#include "absl/memory/memory.h"
+
+namespace util = firebase::firestore::util;
+using firebase::firestore::auth::CredentialsProvider;
+using firebase::firestore::auth::FirebaseCredentialsProvider;
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface FSTFirestoreComponent () <FIRComponentLifecycleMaintainer, FIRComponentRegistrant>
+@end
+
+@implementation FSTFirestoreComponent
+
+// Explicitly @synthesize because instances is part of the FSTInstanceProvider protocol.
+@synthesize instances = _instances;
+
+#pragma mark - Initialization
+
+- (instancetype)initWithApp:(FIRApp *)app {
+  self = [super init];
+  if (self) {
+    _instances = [[NSMutableDictionary alloc] init];
+
+    HARD_ASSERT(app, "Cannot initialize Firestore with a nil FIRApp.");
+    _app = app;
+  }
+  return self;
+}
+
+#pragma mark - FSTInstanceProvider Conformance
+
+- (FIRFirestore *)firestoreForDatabase:(NSString *)database {
+  if (!database) {
+    FSTThrowInvalidArgument(@"database identifier may not be nil.");
+  }
+
+  NSString *key = [NSString stringWithFormat:@"%@|%@", self.app.name, database];
+
+  // Get the component from the container.
+  @synchronized(self.instances) {
+    FIRFirestore *firestore = _instances[key];
+    if (!firestore) {
+      // Question for Gil: Do we want to key based on app name as well here?
```

Comment:
I'm not opposed, but I think the default app name is fugly, so I'd prefer the default to remain as is. Are you thinking of something like this?

```c++
std::string queue_name{"com.google.firebase.firestore"};
if (self.app.name is not the default app) {
  absl::StrAppend(&queue_name, ".", util::MakeStringView(self.app.name));
}
... dispatch_queue_create(queue_name.c_str(), DISPATCH_QUEUE_SERIAL);
```

Reply:
That sounds good to me, do you foresee any issues with instances being on separate `dispatch_queue`s?

Reply:
Nope. The only thing we care about is that within an instance all operations happen on the same dispatch queue.